base/src/source.rs

```diff
@@ -0,0 +1,189 @@
+use std::fmt;
+use std::fs::File;
+use std::io::{self, Read};
+use std::path::PathBuf;
+
+use pos::{BytePos, CharPos, Location};
+
+#[derive(Clone, Debug)]
+pub enum SourceName {
+    /// Placeholder name for sources that do not originate on the file system
+    Placeholder(String),
+    /// Path to the file that this source code coresponds to
+    FilePath(PathBuf),
+}
+
+impl fmt::Display for SourceName {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            SourceName::Placeholder(ref name) => write!(f, "<{}>", name),
+            SourceName::FilePath(ref name) => name.to_str().expect("Invalid filename").fmt(f),
+        }
```

Comment:
I think its better to use [display](https://doc.rust-lang.org/std/path/struct.Path.html#method.display) here (a method I never heard of before, wonder why it is needed)